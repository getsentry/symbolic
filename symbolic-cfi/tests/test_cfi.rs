use std::str;

use symbolic_cfi::{AsciiCfiWriter, CfiCache};
use symbolic_common::ByteView;
use symbolic_debuginfo::Object;
use symbolic_testutils::fixture;

use similar_asserts::assert_eq;

type Error = Box<dyn std::error::Error>;

#[test]
fn load_empty_cfi_cache() -> Result<(), Error> {
    let buffer = ByteView::from_slice(&[]);
    let cache = CfiCache::from_bytes(buffer)?;
    assert_eq!(cache.version(), 1);
    Ok(())
}

#[test]
fn cfi_from_elf() -> Result<(), Error> {
    let buffer = ByteView::open(fixture("linux/crash"))?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    // NOTE: Breakpad's CFI writer outputs registers in alphabetical order. We
    // write the CFA register first, and then order by register number. Thus,
    // the output is not identical to `cfi_sym_linux`.
    insta::assert_snapshot!("cfi_elf", cfi);

    Ok(())
}

#[test]
fn cfi_from_macho() -> Result<(), Error> {
    let buffer = ByteView::open(fixture("macos/crash"))?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    // NOTE: Breakpad's CFI writer outputs registers in alphabetical order. We
    // write the CFA register first, and then order by register number. Thus,
    // the output is not identical to `cfi_sym_macos`.
    insta::assert_snapshot!("cfi_macho", cfi);

    Ok(())
}

#[test]
fn cfi_from_sym_linux() -> Result<(), Error> {
    let buffer = ByteView::open(fixture("linux/crash.sym"))?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot!("cfi_sym_linux", cfi);

    Ok(())
}

#[test]
fn cfi_from_sym_macos() -> Result<(), Error> {
    let buffer = ByteView::open(fixture("macos/crash.sym"))?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot!("cfi_sym_macos", cfi);

    Ok(())
}

#[test]
fn cfi_from_sym_windows() -> Result<(), Error> {
    let buffer = ByteView::open(fixture("windows/crash.sym"))?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot!("cfi_sym_windows", cfi);

    Ok(())
}

#[test]
fn cfi_from_pdb_windows() -> Result<(), Error> {
    let buffer = ByteView::open(fixture("windows/crash.pdb"))?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot!("cfi_pdb_windows", cfi);

    Ok(())
}

#[test]
fn cfi_from_pe_windows() -> Result<(), Error> {
    let buffer = ByteView::open(fixture("windows/CrashWithException.exe"))?;
    let object = Object::parse(&buffer)?;

    let buf: Vec<u8> = AsciiCfiWriter::transform(&object)?;
    let cfi = str::from_utf8(&buf)?;
    insta::assert_snapshot!("cfi_pe_windows", cfi);

    Ok(())
}
