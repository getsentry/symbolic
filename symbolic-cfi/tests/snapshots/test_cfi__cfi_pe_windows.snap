---
source: symbolic-cfi/tests/test_cfi.rs
assertion_line: 105
expression: cfi
---
STACK CFI INIT 1010 55 .cfa: $rsp 80 + $rbx: .cfa 16 - ^ $rsi: .cfa 24 - ^ $rdi: .cfa 32 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 10b0 43 .cfa: $rsp 48 + $rdi: .cfa 16 - ^ $rbx: .cfa 0 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1120 2d .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1150 18 .cfa: $rsp 64 + .ra: .cfa 8 - ^
STACK CFI INIT 1170 a .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1190 21 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI INIT 11bc 3c .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 11f8 2b .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1224 b7 .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 12dc 10 .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 12ec 19 .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1308 17c .cfa: $rsp 64 + $rdi: .cfa 16 - ^ $rbx: .cfa 0 - ^ $rsi: .cfa -8 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1484 12 .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1498 34 .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 14cc d1 .cfa: $rsp 64 + .ra: .cfa 8 - ^
STACK CFI INIT 15a0 71 .cfa: $rsp 96 + $rbx: .cfa 16 - ^ $rsi: .cfa 24 - ^ $rdi: .cfa 32 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 161c 3f .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 167c 3f .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 16dc 35 .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1728 42 .cfa: $rsp 48 + $rdi: .cfa 16 - ^ $rbx: .cfa 0 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 176c 20 .cfa: $rsp 80 + .ra: .cfa 8 - ^
STACK CFI INIT 178c 20 .cfa: $rsp 80 + .ra: .cfa 8 - ^
STACK CFI INIT 17ac 39 .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 17e8 49 .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1834 d8 .cfa: $rsp 80 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 190c 99 .cfa: $rsp 32 + .ra: .cfa 8 - ^
STACK CFI INIT 19a8 24 .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 19cc 2b .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 19f8 4f .cfa: $rsp 48 + $rbx: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1a48 17 .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1a60 ac .cfa: $rsp 48 + $rbp: .cfa 16 - ^ $rbx: .cfa -24 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1b40 1b .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1b80 14a .cfa: $rsp 1488 + $rbp: .cfa 16 - ^ $rbx: .cfa 0 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1cd4 52 .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1d38 38 .cfa: $rsp 48 + .ra: .cfa 8 - ^
STACK CFI INIT 1d70 3c .cfa: $rsp 48 + $rdi: .cfa 16 - ^ $rbx: .cfa 0 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1dac 3c .cfa: $rsp 48 + $rdi: .cfa 16 - ^ $rbx: .cfa 0 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 1de8 1b9 .cfa: $rsp 48 + $rsi: .cfa 16 - ^ $rdi: .cfa 24 - ^ $r14: .cfa 32 - ^ $rbx: .cfa -8 - ^ $rbp: .cfa -16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 2090 2 .cfa: $rsp 8 + .ra: .cfa 8 - ^
STACK CFI INIT 20ac 1e .cfa: $rsp 48 + $rbp: .cfa 16 - ^ .ra: .cfa 8 - ^
STACK CFI INIT 20ca 18 .cfa: $rsp 16 + $rbp: .cfa 16 - ^ .ra: .cfa 8 - ^

